### DISCUSSION TITLE
Migration engine download in build pipeline

### DISCUSSION THREAD

**Discussion by sparkofreason** [prisma/prisma#5965] [Q&A] [answered]
Trying to integrate with our build pipeline, getting an error when `npx prisma generate` runs and attempts to download

https://binaries.prisma.sh/all_commits/3c463ebd78b1d21d8fdacdd27899e280cf686223/rhel-openssl-1.0.x/migration-engine.gz

This is expected, since our build system does not allow access to external repositories. Is this not available on npm? Is there some recommended approach here? Worst case we can manually add to our artifactory, wondering if there's some other path.

**Comment by sparkofreason** [comment, marked as answer]
Not sure if it matters, but we aren't using Prisma for migrations, so maybe there's some configuration to skip this download?

**Reply by janpio** [reply]
You would get the same message for all other binaries, so this would unfortunately not help here.

**Reply by sparkofreason** [reply]
Indeed, as I discovered ;). The custom binary route feels convoluted as well, because our development and deployment environments are not the same. 

I almost want to just mirror `binaries.prisma.sh`, and be able to redirect the base URL or path, so the environment etc. is correctly resolved when `generate` runs. 

I assume others have this issue, and there is some generally accepted solution?

**Reply by janpio** [reply]
The difference between environments is also the reason why we decided to now put them into an Npm package of course - at the cost of what you describe.

I am not aware what other people are doing, besides what you already suggested yourself. I'll raise this internally.

**Reply by Jolg42** [reply]
We have a dedicated env var if you want to have your own mirror `PRISMA_BINARIES_MIRROR` it defaults to `https://binaries.prisma.sh`

**Comment by sparkofreason** [comment]
I think I found the answer: https://www.prisma.io/docs/concepts/overview/under-the-hood#using-custom-engine-binaries

**Comment by Jolg42** [comment]
Note: There is a related issue here https://github.com/prisma/prisma/issues/5105

**Comment by mwitteveen** [comment]
Did you find a download location in the end?

**Reply by sparkofreason** [reply]
No.

**Reply by Jolg42** [reply]
@mwitteveen What are you trying to do? / What's missing? We can maybe help here or have an issue to improve something 👀

**Reply by mwitteveen** [reply]
@Jolg42  So I am developing a mac m1 max then at my client (a government agency) the whole building is disconnected from the internet. So I need to download the binaries onto a USB stick and go on the server. I can develop on VM's they host for me (those are ubuntu) but still not connected to the internet. Then the backend is running in RHEL thus again another engine is needed. 

It ends up that I have multiple laptops with me to install the @prisma/engine package en then take the binaries from the installed package. For the RHEL binaries now use a Linux laptop I take with me to launch a docker container running RHEL and then again install @prisma/engines to then copy the binaries out of the container and onto a USB stick.

In the cicd pipeline (gitlab ci) I'm running postgres as sidecontainer to test against. For this to work I need to reset that database. Therefore I need the migration engine inside the pipeline. Then because the project is to big I need multiple .schema files. (I still do not understand that this is not supported). For this I am using prismix. But that package is using the format binary. So the binary target option for generating the prisma client is not enough as both the formatting and migration engine are not included.

**Reply by Jolg42** [reply]
@mwitteveen So there are 2 different and complementary ways to influence the downloading of Prisma engines.

For the CLI, this could help:
```
// Delete node_modules first to make sure it works
rm -rf ./node_modules
// Run npm install with the env var
PRISMA_CLI_BINARY_TARGETS=darwin,debian-openssl-1.1.x,rhel-openssl-1.1.x npm install
```
https://www.prisma.io/docs/reference/api-reference/environment-variables-reference#cli-binary-targets

For the generator clock (for Prisma Client's Query Engine)
```
generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["darwin", "debian-openssl-1.1.x", "rhel-openssl-1.1.x"]
}
```
https://www.prisma.io/docs/concepts/components/prisma-schema/generators#binary-targets

I think this could be improved though to better serve offline/secure setup like yours.
Let me know if that helped 👀 

Note: you'll need to pick the right `openssl` version between `1.0.x`, `1.1.x` and `3.0.x`
https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#binarytargets-options

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
